### PR TITLE
[Java] Add Sequence Recognizer

### PR DESCRIPTION
**Pull request by Batta32**
## Description
The Sequence Recognizer is missing in Java. This PR contains the migrated recognizer from C# to Java (only code).

This PR contains the English parsers/extractors.

## Specific Changes
- Add main structure of the `recognizers-sequence`
- Add recognizer as submodule of `recognizers-text`
- Migrate English Extractors & Parsers
- Migrate root files
- Integrate Sequence in Simple console
- Overload [trimStart](https://github.com/southworks/Recognizers-Text/blob/master/Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/utilities/StringUtility.java#L12) of `StringUtility`
- Catch infinite values in [NumberFormatUtility](https://github.com/southworks/Recognizers-Text/blob/master/Java/libraries/recognizers-text-number/src/main/java/com/microsoft/recognizers/text/number/parsers/NumberFormatUtility.java#L33)

## Testing
We created PR #2466 which contains the unit tests passing correctly (524/524) following the C# & JS unit tests.

_Migrated structure of Sequence_
image

image

image

### CODE REVIEW

**Review by tellarin — CHANGES_REQUESTED** [review]
Build is not passing. Can you see if you can figure out why? It’s not clear from the log and re-triggering the build didn’t solve it.

**Comment by Batta32** [comment]
Hi @tellarin, I was reviewing the [last build](https://ci.appveyor.com/project/tellarin/recognizers-text/builds/37652515) and apparently it's failing in `maven-checkstyle-plugin:3.0.0:check (validate) @ recognizers-text-sequence`, however, I couldn't replicate it locally using different Java versions.

Also, I confirm that I'm using the `linting-rules.xml` as rules of Checkstyle.

Can it be the configuration of the CI?

**Comment by Batta32** [comment]
@tellarin, we executed the `build.cmd` of Java locally and we noticed that the build failed because the resources of sequence were not generated. 

We will be double-checking this and update this PR if it's needed.

image

**Comment by Batta32** [comment]
Hi @tellarin, we successfully **fixed** the build of this PR.

We applied the following changes:
- Include the `resource-definitions.json` of Sequence in the `pom.xml` of the `resource-generator` for the autogenerated resources during the build.
- Update the `resource-definitions.json` of Sequence to follow the approach of [Python](https://github.com/microsoft/Recognizers-Text/blob/master/Python/libraries/recognizers-sequence/resource-definitions.json#L123) and [JS](https://github.com/microsoft/Recognizers-Text/blob/master/JavaScript/packages/recognizers-sequence/resource-definitions.json#L92) to have the `PhoneNumbersDefinitions` as alias of `ChinesePhoneNumbers`
- Remove redundant `BasicIpv6Element` of [Chinese-Ip.yaml](https://github.com/microsoft/Recognizers-Text/blob/master/Patterns/Chinese/Chinese-Ip.yaml#L7)

File: Java/libraries/recognizers-text-sequence/src/main/java/com/microsoft/recognizers/text/sequence/extractors/BaseEmailExtractor.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        return results.stream().filter((o -> RFC_5322_VALIDATION_REGEX.matcher((o).getText()).matches()))
+                .collect(Collectors.toList());
+    }
+
+    public BaseEmailExtractor(BaseSequenceConfiguration config) {
+        this.config = config;
+        Map<Pattern, String> regexes = new HashMap<Pattern, String>();
+        regexes.put(Pattern.compile(BaseEmail.EmailRegex), Constants.EMAIL_REGEX);
+        // EmailRegex2 will break the code as it's not supported in Java, comment out for now
+        // regexes.put(Pattern.compile(BaseEmail.EmailRegex2), Constants.EMAIL_REGEX);
```
Comment (tellarin) [inline Java/libraries/recognizers-text-sequence/src/main/java/com/microsoft/recognizers/text/sequence/extractors/BaseEmailExtractor.java:57]: Why is it not supported in Java? What are the side-effects of commenting this out?
In such cases, please add a `@TODO` comment to implement a solution or workaround.

File: Java/libraries/recognizers-text-sequence/src/main/java/com/microsoft/recognizers/text/sequence/spanish/SpanishPhoneNumberExtractorConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.recognizers.text.sequence.spanish;
```
Comment (tellarin) [inline Java/libraries/recognizers-text-sequence/src/main/java/com/microsoft/recognizers/text/sequence/spanish/SpanishPhoneNumberExtractorConfiguration.java:4]: Do we need this? If I recall correctly, all non-CJK languages shoud fallback to the English PhoneNumber model as all formats should be supported in any language.

File: Java/samples/simple-console/src/main/java/com/microsoft/recognizers/text/samples/simpleconsole/Sample.java [hunk 3]
```diff
@@ -94,7 +96,35 @@ public static void main(String[] args) throws Exception {
 
                 // Datetime recognizer This model will find any Date even if its write in colloquial language
                 // E.g "I'll go back 8pm today" will return "2017-10-04 20:00:00"
-                DateTimeRecognizer.recognizeDateTime(query, culture)
+                DateTimeRecognizer.recognizeDateTime(query, culture),
+
+                // PhoneNumber recognizer will find any phone number presented
+                // E.g "My phone number is ( 19 ) 38294427."
+                SequenceRecognizer.recognizePhoneNumber(query, culture, SequenceOptions.None, true),
```
Comment (tellarin) [inline Java/samples/simple-console/src/main/java/com/microsoft/recognizers/text/samples/simpleconsole/Sample.java:103]: Similarly to the other models, the last two parameters should not be required, unless a user wants to override the defaults.

**Review by tellarin — CHANGES_REQUESTED** [review]

File: Java/libraries/recognizers-text-sequence/src/main/java/com/microsoft/recognizers/text/sequence/spanish/SpanishPhoneNumberExtractorConfiguration.java (hunk 2 above)
Comment (Batta32) [reply Java/libraries/recognizers-text-sequence/src/main/java/com/microsoft/recognizers/text/sequence/spanish/SpanishPhoneNumberExtractorConfiguration.java]: We removed the `SpanishPhoneNumberExtractorConfiguration` as you mentioned.

File: Java/samples/simple-console/src/main/java/com/microsoft/recognizers/text/samples/simpleconsole/Sample.java (hunk 3 above)
Comment (Batta32) [reply Java/samples/simple-console/src/main/java/com/microsoft/recognizers/text/samples/simpleconsole/Sample.java]: Done!

File: Java/libraries/recognizers-text-sequence/src/main/java/com/microsoft/recognizers/text/sequence/extractors/BaseEmailExtractor.java (hunk 1 above)
Comment (Batta32) [reply Java/libraries/recognizers-text-sequence/src/main/java/com/microsoft/recognizers/text/sequence/extractors/BaseEmailExtractor.java:61]: We [added](https://github.com/microsoft/Recognizers-Text/pull/2465/commits/78e2353a0814d0f2c345079381805d899ae2368b) a `@TODO` comment. The approach was the same as JavaScript which was added in PR #905 and also like Python in PR #1489 (see [comment](https://github.com/microsoft/Recognizers-Text/pull/1489#discussion_r264083539))

_Exception is EmailRegex2 is commented out_
image

**Review by tellarin — APPROVED** [review]

**Pull request merged**